lua/cfc_notifications/client/cl_save.lua

```diff
@@ -0,0 +1,157 @@
+-- Save/load notification perferences ( such as "ignore this session" and "ignore forever" )
+CFCNotifications._tempIgnores = CFCNotifications._tempIgnores or {}
+CFCNotifications._permIgnores = {}
+CFCNotifications._SAVE_FILE_NAME = "cfc_notifications_data.json"
+
+-- Add some way to browse ignores in settings and call Unignore on them
+
+function CFCNotifications.Base:ShouldShowNotification()
+    local id = self:GetID()
+    return not CFCNotifications._tempIgnores[id] and not CFCNotifications._permIgnores[id]
+end
+
+function CFCNotifications.Base:Ignore( permanent )
+    local id = self:GetID()
+
+    if permanent then
+        CFCNotifications._permIgnores[id] = true
+        CFCNotifications.saveIgnores()
+    else
+        CFCNotifications._tempIgnores[id] = true
+    end
+    CFCNotifications._reloadIgnoredPanels()
+end
+
+function CFCNotifications.Base:Unignore()
+    local id = self:GetID()
+
+    CFCNotifications._tempIgnores[id] = nil
+    if CFCNotifications._permIgnores[id] then
+        CFCNotifications._permIgnores[id] = nil
+        CFCNotifications.saveIgnores()
+    end
+    CFCNotifications._reloadIgnoredPanels()
+end
+
+function CFCNotifications.saveIgnores()
+    local strData = util.TableToJSON( table.GetKeys( CFCNotifications._permIgnores ) )
+    file.Write( CFCNotifications._SAVE_FILE_NAME, strData )
+end
+
+function CFCNotifications.loadIgnores()
+    local strData = file.Read( CFCNotifications._SAVE_FILE_NAME )
+    if not strData then return end
+    local data = util.JSONToTable( strData )
+    for k = 1, #data do
+        CFCNotifications._permIgnores[data[k]] = true
+    end
+    CFCNotifications._reloadIgnoredPanels()
+end
+
+function CFCNotifications.getTempIgnored()
+    return table.GetKeys( CFCNotifications._tempIgnores )
+end
+
+function CFCNotifications.getPermIgnored()
+    return table.GetKeys( CFCNotifications._permIgnores )
+end
+
+function CFCNotifications.getUnignored()
+    local out = {}
+    for k, v in pairs( CFCNotifications.Notifications ) do
+        if not CFCNotifications._tempIgnores[k] and not CFCNotifications._permIgnores[k] and v:GetCloseable() and v:GetIgnoreable() then
+            table.insert( out, k )
+        end
+    end
+    for k, v in pairs( CFCNotifications._serverNotificationIDs ) do
+        if not CFCNotifications._tempIgnores[v] and not CFCNotifications._permIgnores[v] and not table.HasValue( out, v ) then
+            table.insert( out, v )
+        end
+    end
+    return out
+end
```

Comment:
Can we get some whitespace in here?

```suggestion
function CFCNotifications.getUnignored()
    local out = {}
    
    for k, v in pairs( CFCNotifications.Notifications ) do
        if not CFCNotifications._tempIgnores[k] and not CFCNotifications._permIgnores[k] and v:GetCloseable() and v:GetIgnoreable() then
            table.insert( out, k )
        end
    end
    
    for k, v in pairs( CFCNotifications._serverNotificationIDs ) do
        if not CFCNotifications._tempIgnores[v] and not CFCNotifications._permIgnores[v] and not table.HasValue( out, v ) then
            table.insert( out, v )
        end
    end
    
    return out
end
```